docs/source/federation-spec.mdx

```diff
@@ -0,0 +1,625 @@
+---
+title: Apollo Federation subgraph specification
+description: For adding subgraph support to GraphQL server libraries
+---
+
+> This content is provided for developers adding federated subgraph support to a GraphQL server library, and for anyone curious about the inner workings of federation. You do _not_ need to read this if you're building a supergraph with existing [subgraph-compatible libraries](./supported-subgraphs/), such as Apollo Server.
+>
+> Servers that are partially or fully compatible with this specification are tracked in Apollo's [subgraph compatibility repository](https://github.com/apollographql/apollo-federation-subgraph-compatibility).
+
+For a GraphQL service to operate as an Apollo Federation 2 subgraph, it must do all of the following:
+
+- Automatically extend its schema with all definitions listed in the [subgraph schema specification](#subgraph-schema-specification)
+- Correctly resolve the `Query._service` [enhanced introspection field](#fetch-service-capabilities)
+- Correctly generate [stubs for externally referenced types](#generating-stub-types)
+- Correctly resolve requests for entity fields via the [`Query._entities` field](#resolving-entity-fields-with-query_entities)
+
+Each of these requirements is described in the sections below.
+
+## Subgraph schema specification
+
+A subgraph must automatically add all of the following definitions to its GraphQL schema. The purpose of each definition is described in [Glossary of schema additions](#glossary-of-schema-additions).
+
+```graphql {1-4}
+# ⚠️ This definition must be created dynamically! The union
+#   must include every object type in the schema that uses
+#   the @key directive (i.e., all federated entities).
+union _Entity
+
+scalar _Any
+scalar FieldSet
+scalar link__Import
```

Comment:
technically `link__` namespace could be skipped IFF those elements are imported through `@link` directive, similarly how we skip the default `federation__` namespace for all other directives.

If elements are not imported through `@link` they need to be namespaced either with default value from the linked spec OR based on the specific namespace value (`@link(url: <spec url>, as: "myNamespace")`)

Reply:
Should the subgraph spec _require_ this definition alongside the other required definitions?

```
extend schema
  @link(url: "https://specs.apollo.dev/link/v1.0/", import: ["@link", "Import", "Purpose"])
```

Not necessarily because it's necessary to un-namespace `Import` and `Purpose`, but because otherwise `@link` isn't being tied to its associated spec

Reply:
I believe the current suggestion is to NOT import the link spec by default as it is automatically added by the router. Devs can import the link spec if they want to avoid name-spacing of `Import` scalar and `Purpose` enum, but should omit it otherwise.

Side note: when importing a spec, it automatically pulls in the corresponding directive matching the spec - so we don't need to explicitly import `@link` in the example above (by importing federation spec we also pull in `@federation` directive).